src/spec-configuration/containerFeaturesOrder.ts

```diff
@@ -0,0 +1,108 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+
+import { ContainerError } from '../spec-common/errors';
+import { FeatureSet } from '../spec-configuration/containerFeaturesConfiguration';
+import { DevContainerConfig } from './configuration';
+
+interface FeatureNode {
+    feature: FeatureSet;
+    before: Set<FeatureNode>;
+    after: Set<FeatureNode>;
+}
+
+export function computeFeatureInstallationOrder(config: DevContainerConfig, features: FeatureSet[]) {
+
+    if (!config.overrideFeatureInstallOrder) {
+        return computeInstallationOrder(features);
+    }
+    else {
+        return computeOverrideInstallationOrder(config, features);
+    }
+}
+
+function computeOverrideInstallationOrder(config: DevContainerConfig, features: FeatureSet[]) {
+    const automaticOrder = computeInstallationOrder(features);
+
+    const orderedFeatures = [];
+
+    for (const featureId of config.overrideFeatureInstallOrder!) {
+        const feature = automaticOrder.find(feature => feature.features[0].name === featureId);
+        if (!feature) {
+            throw new ContainerError({ description: `Feature ${featureId} not found` });
+        }
+        orderedFeatures.push(feature);
+        features.splice(features.indexOf(feature), 1);
+    }
+
+    return orderedFeatures.concat(features);
+}
+
+function computeInstallationOrder(features: FeatureSet[]) {
```

Comment:
I think a comment on this function illustrating the set of rules its following would be helpful

eg:

- Looks at each feature's `installAfter` to resolve soft dependencies
- ???